flow/Net2.actor.cpp

```diff
@@ -487,6 +509,183 @@ class Connection : public IConnection, ReferenceCounted<Connection> {
+		p.getEndpoint() = std::make_shared<udp::endpoint>();
+		auto endpoint = p.getEndpoint().get();
+		auto res = p.getFuture();
+		socket.async_receive_from(boost::asio::mutable_buffer(begin, end - begin), *endpoint, std::move(p));
+		return fmap(
+		    [endpoint, sender](int bytes) {
+			    if (sender) {
+				    sender->port = endpoint->port();
+				    sender->ip = toIPAddress(endpoint->address());
+			    }
+			    g_net2->udpBytesReceived += bytes;
+			    return bytes;
+		    },
+		    res);
+	}
+
+	Future<int> send(uint8_t const* begin, uint8_t const* end) override {
+		++g_net2->countUDPWrites;
+		ReadPromise p("N2_UDPWriteError", id);
+		auto res = p.getFuture();
+		socket.async_send(boost::asio::const_buffer(begin, end - begin), std::move(p));
+		return res;
+	}
+
+	Future<int> sendTo(uint8_t const* begin, uint8_t const* end, NetworkAddress const& peer) override {
+		++g_net2->countUDPWrites;
+		ReadPromise p("N2_UDPWriteError", id);
+		auto res = p.getFuture();
+		udp::endpoint toEndpoint = udpEndpoint(peer);
+		socket.async_send_to(boost::asio::const_buffer(begin, end - begin), toEndpoint, std::move(p));
+		return res;
+	}
+
+	void bind(NetworkAddress const& addr) override {
+		boost::system::error_code ec;
+		socket.bind(udpEndpoint(addr), ec);
+		if (ec) {
+			Error x;
+			if (ec.value() == EADDRINUSE)
+				x = address_in_use();
+			else if (ec.value() == EADDRNOTAVAIL)
+				x = invalid_local_address();
+			else
+				x = bind_failed();
+			TraceEvent(SevWarnAlways, "Net2UDPBindError").error(x);
+			throw x;
+		}
+		isPublic = true;
+	}
+
+	UID getDebugID() const override { return id; }
+
+	void addref() override { ReferenceCounted<UDPSocket>::addref(); }
+	void delref() override { ReferenceCounted<UDPSocket>::delref(); }
+
+	NetworkAddress localAddress() const {
+		auto endpoint = socket.local_endpoint();
+		return NetworkAddress(toIPAddress(endpoint.address()), endpoint.port(), isPublic, false);
+	}
+
+private:
+	UDPSocket(boost::asio::io_service& io_service, Optional<NetworkAddress> toAddress, bool isV6)
+	  : id(nondeterministicRandom()->randomUniqueID()), socket(io_service, isV6 ? udp::v6() : udp::v4()) {
+		socket.non_blocking(true);
+	}
+
+	void closeSocket() {
+		boost::system::error_code error;
+		socket.close(error);
+		if (error)
+			TraceEvent(SevWarn, "N2_CloseError", id)
+			    .suppressFor(1.0)
+			    .detail("ErrorCode", error.value())
+			    .detail("Message", error.message());
+	}
+
+	void onReadError(const boost::system::error_code& error) {
+		TraceEvent(SevWarn, "N2_UDPReadError", id)
+		    .suppressFor(1.0)
+		    .detail("ErrorCode", error.value())
+		    .detail("Message", error.message());
+		closeSocket();
+	}
+	void onWriteError(const boost::system::error_code& error) {
+		TraceEvent(SevWarn, "N2_UDPWriteError", id)
+		    .suppressFor(1.0)
+		    .detail("ErrorCode", error.value())
+		    .detail("Message", error.message());
+		closeSocket();
+	}
+
+	void init() {
+		socket.non_blocking(true);
```

Comment:
Is there a reason to set `socket` to non-blocking here if it's already set in the constructor?

Reply:
that was a refactoring mistake..